pkg/sessionctx/variable/session.go

```diff
@@ -254,6 +254,9 @@ type TxnCtxNoNeedToRestore struct {
 	// Read results cannot be directly written into pessimisticLockCache because failed statement need to rollback
 	// its pessimistic locks.
 	CurrentStmtPessimisticLockCache map[string][]byte
+
+	// MemBufferSnapshot is a snapshot of the memory buffer, every in-txn read of the buffer should use the snapshot to bypass the mutations of current statement.
+	MemBufferSnapshot kv.MemBufferSnapshot
```

Comment:
It seems more like a statement-level information. Should it be placed in `stmtCtx` and initialized `OnStmtStart`?